File: P4/java/deepspace/src/deepspace/Damage.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

public abstract class Damage {
    protected int nShields;
    
    Damage(int s) {
        nShields = s;
    }
    
    abstract Damage newCopy();
    
    public int getNShields() {
        return nShields;
    }
    
    abstract DamageToUI getUIversion();
    
    protected int adjustNShields(ArrayList<ShieldBooster> s) {
        return nShields < s.size() ? nShields : s.size();
    }
    
    abstract Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s);
    
    public void discardShieldBooster() {
        if (nShields > 0)
            nShields -= 1;
    }
        
    abstract void discardWeapon(Weapon w);
    
    public boolean hasNoEffect() {
        return nShields == 0;
    }
}

File: P4/ruby/Damage.rb
# encoding: UTF-8 

require_relative 'DamageToUI'

module Deepspace 
    class Damage
        def initialize (s)
            @nShields = s # : int
        end

        attr_reader :nShields

        def adjustNShields (s) # adjustNShields (s: Array<ShieldBooster>) : int
            if s != nil
            	[@nShields, s.length].min
           	end
        end

        def discardShieldBooster
            if(@nShields > 0)
                @nShields -= 1
            end
        end

        def hasNoEffect
        	@nShields == 0
        end

        # Especificación de acceso a métodos
        protected :adjustNShields
    end
end
